File: Module-3-TP-Patterns/python/main/main.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from models import Etudiant
from singleton import ScolariteManager
from factory import PersonneFactory
from adapter import LegacyCoursSystem, CoursAdapter
from decorator import EtudiantBoursier, EtudiantDelegue
from strategy import TriParMoyenne, TriParNom

print("=== SYSTEME DE GESTION ACADEMIQUE ===\n")

# 1. SINGLETON
print("1. SINGLETON")
manager = ScolariteManager()
print(f"   Meme instance ? {manager is ScolariteManager()}")

# 2. FACTORY METHOD
print("\n2. FACTORY METHOD")
alice   = PersonneFactory.get_factory("etudiant").creer("Alice",   20, "E001")
bob     = PersonneFactory.get_factory("etudiant").creer("Bob",     22, "E002")
charlie = PersonneFactory.get_factory("etudiant").creer("Charlie", 21, "E003")
prof    = PersonneFactory.get_factory("professeur").creer("Prof. Martin", 45, "Informatique")
alice.afficher()
prof.afficher()

# 4. ADAPTER
print("\n4. ADAPTER")
cours = CoursAdapter(LegacyCoursSystem()).get_all_cours()
for c in cours:
    print(f"   Importe : {c}")

alice.ajouter_cours(cours[0]); alice.ajouter_cours(cours[1])
bob.ajouter_cours(cours[0]);   charlie.ajouter_cours(cours[2])

manager.ajouter_etudiant(alice)
manager.ajouter_etudiant(bob)
manager.ajouter_etudiant(charlie)

# 6. OBSERVER
print("\n6. OBSERVER")
alice.ajouter_note(15.5)
bob.ajouter_note(12.0)
charlie.ajouter_note(18.0)

# 3. DECORATOR
print("\n3. DECORATOR")
EtudiantBoursier(alice, 500.0).afficher()
print()
EtudiantDelegue(bob, "L3 INFO").afficher()

# 5. STRATEGY
print("\n5. STRATEGY - tri par moyenne :")
manager.set_tri_strategy(TriParMoyenne())
for e in manager.get_etudiants_tries():
    print(f"   {e.nom:<10} -> {e.moyenne}")

print("   Changement -> tri par nom :")
manager.set_tri_strategy(TriParNom())
for e in manager.get_etudiants_tries():
    print(f"   {e.nom:<10} -> {e.moyenne}")

File: Module-3-TP-Patterns/java/main/Main.java
public class Main {
    public static void main(String[] args) {

        System.out.println("=== SYSTEME DE GESTION ACADEMIQUE ===\n");

        // 1. SINGLETON
        System.out.println("1. SINGLETON");
        ScolariteManager manager = ScolariteManager.getInstance();
        System.out.println("   Meme instance ? " + (manager == ScolariteManager.getInstance()));

        // 2. FACTORY METHOD
        System.out.println("\n2. FACTORY METHOD");
        Etudiant alice   = (Etudiant)   PersonneFactory.getFactory("etudiant").creer("Alice",   20, "E001");
        Etudiant bob     = (Etudiant)   PersonneFactory.getFactory("etudiant").creer("Bob",     22, "E002");
        Etudiant charlie = (Etudiant)   PersonneFactory.getFactory("etudiant").creer("Charlie", 21, "E003");
        Professeur prof  = (Professeur) PersonneFactory.getFactory("professeur").creer("Prof. Martin", 45, "Informatique");
        alice.afficher();
        prof.afficher();

        // 4. ADAPTER
        System.out.println("\n4. ADAPTER");
        Cours[] cours = new CoursAdapter(new LegacyCoursSystem()).getAllCours();
        for (Cours c : cours) System.out.println("   Importe : " + c);

        alice.ajouterCours(cours[0]); alice.ajouterCours(cours[1]);
        bob.ajouterCours(cours[0]);   charlie.ajouterCours(cours[2]);

        manager.ajouterEtudiant(alice);
        manager.ajouterEtudiant(bob);
        manager.ajouterEtudiant(charlie);

        // 6. OBSERVER
        System.out.println("\n6. OBSERVER");
        alice.ajouterNote(15.5);
        bob.ajouterNote(12.0);
        charlie.ajouterNote(18.0);

        // 3. DECORATOR
        System.out.println("\n3. DECORATOR");
        new EtudiantBoursier(alice, 500.0).afficher();
        System.out.println();
        new EtudiantDelegue(bob, "L3 INFO").afficher();

        // 5. STRATEGY
        System.out.println("\n5. STRATEGY - tri par moyenne :");
        manager.setTriStrategy(new TriParMoyenne());
        manager.getEtudiantsTries().forEach(e ->
            System.out.printf("   %-10s -> %.1f%n", e.getNom(), e.getMoyenne()));

        System.out.println("   Changement -> tri par nom :");
        manager.setTriStrategy(new TriParNom());
        manager.getEtudiantsTries().forEach(e ->
            System.out.printf("   %-10s -> %.1f%n", e.getNom(), e.getMoyenne()));
    }
}
